src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/DotnetCliHelper.cs

```diff
@@ -0,0 +1,172 @@
+    {
+        _logger = loggerFactory.CreateLogger<DotnetCliHelper>();
+        _dotnetExecutablePath = new Lazy<string>(() => GetDotnetExecutablePath());
+        _dotnetSdkFolder = new AsyncLazy<string>(GetDotnetSdkFolderFromDotnetExecutableAsync, cacheResult: true);
+    }
+
+    public string GetDotnetExecutablePath()
+    {
+        // The client can modify the value in DOTNET_ROOT to ensure that the server starts on the correct runtime.
+        // It will save the user's DOTNET_ROOT in the DOTNET_ROOT_USER environment variable, so check there instead of using the DOTNET_ROOT.
+        var dotnetPathFromDotnetRootUser = Environment.GetEnvironmentVariable(DotnetRootUserEnvVar);
+        if (TryGetDotnetExecutableFromFolder(dotnetPathFromDotnetRootUser, DotnetRootUserEnvVar, _logger, out var dotnetRootUserExecutable))
+        {
+            return dotnetRootUserExecutable;
+        }
+
+        // Neither an option or env var was provided, find the dotnet path using what is currently on the path.
+        _logger.LogInformation("Using dotnet executable configured on the PATH");
+        return "dotnet";
+
+        static bool TryGetDotnetExecutableFromFolder(string? dotnetFolder, string optionName, ILogger logger, [NotNullWhen(true)] out string? dotnetExecutablePath)
+        {
+            if (string.IsNullOrEmpty(dotnetFolder))
+            {
+                dotnetExecutablePath = null;
+                return false;
+            }
+
+            var dotnetExecutableName = $"dotnet{(RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? ".exe" : string.Empty)}";
+            var executablePath = Path.Combine(dotnetFolder, dotnetExecutableName);
+            if (File.Exists(executablePath))
+            {
+                logger.LogInformation($"Found dotnet executable at {executablePath} using {optionName}");
+                dotnetExecutablePath = executablePath;
+                return true;
+            }
+            else
+            {
+                logger.LogInformation($"The {optionName} option {dotnetFolder} does not contain a valid dotnet executable");
+                dotnetExecutablePath = null;
+                return false;
+            }
+        }
+    }
+
+    /// <summary>
+    /// The folder the dotnet executable is in could contain multiple SDK paths.
+    /// In order to figure out which one is the right one, we need to run dotnet --info
+    /// which will output the correct SDK path.
+    /// </summary>
+    private async Task<string> GetDotnetSdkFolderFromDotnetExecutableAsync(CancellationToken cancellationToken)
+    {
+        using var process = Run("--info", workingDirectory: null, shouldLocalizeOutput: false);
+
+        string? dotnetSdkFolderPath = null;
+        process.OutputDataReceived += (_, e) =>
+        {
+            if (!string.IsNullOrEmpty(e.Data))
+            {
+                var line = e.Data.Trim();
+                if (line.StartsWith("Base Path", StringComparison.OrdinalIgnoreCase))
```

Comment:
does this work if the user is working *not* in english?

Reply:
any chance this could happen multiple times?

Reply:
> does this work if the user is working not in english?

Yep - we pass `shouldLocalizeOutput: false` so we will always get back english.

> any chance this could happen multiple times?

Nope - this is a lazy created in the ctor.